rmw_dds_common/include/rmw_dds_common/qos.hpp

```diff
@@ -224,6 +226,28 @@ RMW_DDS_COMMON_PUBLIC
 rmw_qos_profile_t
 qos_profile_update_best_available_for_services(const rmw_qos_profile_t & qos_profile);
 
+/// Parse USER_DATA key=value;key=value; encoding, find value of key "typehash"
+/**
+ * \param[in] user_data USER_DATA qos raw bytes
+ * \param[in] user_data_size Length of user_data
+ * \return Parsed type hash
+ * \return zero-initialized value if data unparseable or key not found
+ */
+RMW_DDS_COMMON_PUBLIC
+rosidl_type_hash_t
+parse_type_hash_from_user_data(
+  const uint8_t * user_data,
+  size_t user_data_size);
+
+/// Encode type hash as "typehash=hash_string;"
+/**
+ * \param[in] type_hash Type hash value to encode
+ * \return A substring in key=value; format that can be appended to QoS USER_DATA
```

Comment:
From what I can tell from the implementation, an empty string may be returned here if the hash is unset, *or* if there is an error.  I think we definitely need to document that, and maybe also consider a way to disambiguate those two cases.

Reply:
Yeah, I wrestled with this a bit. I've put up a version that returns `rmw_ret_t` and takes a `std::string & out` argument.